src/java/main/org/apache/zookeeper/common/ZKConfig.java

```diff
@@ -0,0 +1,230 @@
+    public static final String JGSS_NATIVE = "sun.security.jgss.native";
+
+    private Map<String, String> properties = new HashMap<String, String>();
+
+    /**
+     * properties, which are common to both client and server, are initialized
+     * from system properties
+     */
+    public ZKConfig() {
+        init();
+    }
+
+    /**
+     * @param configPath
+     *            Configuration file path
+     * @throws ConfigException
+     *             if failed to load configuration properties
+     */
+
+    public ZKConfig(String configPath) throws ConfigException {
+        this(new File(configPath));
+    }
+
+    /**
+     *
+     * @param configFile
+     *            Configuration file
+     * @throws ConfigException
+     *             if failed to load configuration properties
+     */
+    public ZKConfig(File configFile) throws ConfigException {
+        init();
+        addConfiguration(configFile);
+    }
+
+    private void init() {
+        /**
+         * backward compatibility for all currently available client properties
+         */
+        handleBackwardCompatibility();
+    }
+
+    /**
+     * Now onwards client code will use properties from this class but older
+     * clients still be setting properties through system properties. So to make
+     * this change backward compatible we should set old system properties in
+     * this configuration.
+     */
+    protected void handleBackwardCompatibility() {
+        properties.put(SSL_KEYSTORE_LOCATION, System.getProperty(SSL_KEYSTORE_LOCATION));
+        properties.put(SSL_KEYSTORE_PASSWD, System.getProperty(SSL_KEYSTORE_PASSWD));
+        properties.put(SSL_TRUSTSTORE_LOCATION, System.getProperty(SSL_TRUSTSTORE_LOCATION));
+        properties.put(SSL_TRUSTSTORE_PASSWD, System.getProperty(SSL_TRUSTSTORE_PASSWD));
+        properties.put(SSL_AUTHPROVIDER, System.getProperty(SSL_AUTHPROVIDER));
+        properties.put(JUTE_MAXBUFFER, System.getProperty(JUTE_MAXBUFFER));
+        properties.put(KINIT_COMMAND, System.getProperty(KINIT_COMMAND));
+        properties.put(JGSS_NATIVE, System.getProperty(JGSS_NATIVE));
+    }
+
+    /**
+     * Get the property value
+     *
+     * @param name
+     * @return property value
+     */
+    public String getProperty(String name) {
+        return properties.get(name);
+    }
+
+    /**
+     * Get the property value, if it is null return default value
+     *
+     * @param name
+     *            Property name
+     * @param defaultValue
+     * @return
+     */
+    public String getProperty(String name, String defaultValue) {
+        String value = properties.get(name);
+        if (value == null) {
+            return defaultValue;
+        } else {
+            return value;
+        }
+    }
+
+    /**
+     * Return the value of "java.security.auth.login.config" system property
+     *
+     * @return value
+     */
+    public String getJaasConfKey() {
```

Comment:
@fpj @arshadmohammad *IMHO*, this class should be 'beefed up' with more methods like this. So that we can properly encapsulate the properties. For example, instead of

```
assertEquals(conf.getProperty(ENABLE_CLIENT_SASL_KEY), "true");
assertEquals(conf.getProperty(ZK_SASL_CLIENT_USERNAME), "ZK");
assertEquals(conf.getProperty(LOGIN_CONTEXT_NAME_KEY), "MyClient");
```

We would have:

```
assertTrue(conf.isEnabledClientSASLKey());
assertEquals(conf.getZkSASLClientUsername(), "ZK");
assertEquals(conf.getLoginContextNameKey(), "MyClient");
```

**OTOH**, I understand that a change like that would be very impactful and potentially a source of bugs. Therefore, something to be done with extreme care, if done at all.